GVFS/GVFS.FunctionalTests/Tests/EnlistmentPerFixture/IncludedFolderTests.cs

```diff
@@ -0,0 +1,262 @@
+            this.ValidateIncludedFolders(childPath);
+
+            this.gvfsProcess.AddIncludedFolders(this.mainIncludedFolder);
+            directories = Directory.GetDirectories(Path.Combine(this.Enlistment.RepoRoot, this.mainIncludedFolder));
+            directories.ShouldMatchInOrder(this.directoriesInMainFolder);
+            this.ValidateIncludedFolders(childPath, this.mainIncludedFolder);
+        }
+
+        [TestCase, Order(3)]
+        public void AddingSiblingFolderShouldNotMakeParentRecursive()
+        {
+            this.gvfsProcess.AddIncludedFolders(this.mainIncludedFolder);
+            this.ValidateIncludedFolders(this.mainIncludedFolder);
+
+            // Add and remove sibling folder to main folder
+            string siblingPath = Path.Combine("GVFS", "FastFetch");
+            this.gvfsProcess.AddIncludedFolders(siblingPath);
+            string folder = this.Enlistment.GetVirtualPathTo(siblingPath);
+            folder.ShouldBeADirectory(this.fileSystem);
+            this.ValidateIncludedFolders(this.mainIncludedFolder, siblingPath);
+
+            this.gvfsProcess.RemoveIncludedFolders(siblingPath);
+            folder.ShouldNotExistOnDisk(this.fileSystem);
+            folder = this.Enlistment.GetVirtualPathTo(this.mainIncludedFolder);
+            folder.ShouldBeADirectory(this.fileSystem);
+            this.ValidateIncludedFolders(this.mainIncludedFolder);
+        }
+
+        [TestCase, Order(4)]
+        public void AddingSubfolderShouldKeepParentRecursive()
+        {
+            this.gvfsProcess.AddIncludedFolders(this.mainIncludedFolder);
+            this.ValidateIncludedFolders(this.mainIncludedFolder);
+
+            // Add subfolder of main folder and make sure it stays recursive
+            string subFolder = Path.Combine(this.mainIncludedFolder, "Properties");
+            this.gvfsProcess.AddIncludedFolders(subFolder);
+            string folder = this.Enlistment.GetVirtualPathTo(subFolder);
+            folder.ShouldBeADirectory(this.fileSystem);
+            this.ValidateIncludedFolders(this.mainIncludedFolder, subFolder);
+
+            folder = this.Enlistment.GetVirtualPathTo(this.mainIncludedFolder, "CommandLine");
+            folder.ShouldBeADirectory(this.fileSystem);
+        }
+
+        [TestCase, Order(5)]
+        [Category(Categories.WindowsOnly)]
+        public void CreatingFolderShouldAddToIncludedListAndStartProjecting()
+        {
+            this.gvfsProcess.AddIncludedFolders(this.mainIncludedFolder);
+            this.ValidateIncludedFolders(this.mainIncludedFolder);
+
+            string newFolderPath = Path.Combine(this.Enlistment.RepoRoot, "GVFS", "GVFS.Common");
+            newFolderPath.ShouldNotExistOnDisk(this.fileSystem);
+            Directory.CreateDirectory(newFolderPath);
+            newFolderPath.ShouldBeADirectory(this.fileSystem);
+            string[] fileSystemEntries = Directory.GetFileSystemEntries(newFolderPath);

```

Comment:
Do we have options on how to handle the case where a user tries to create a directory that is outside the scope of included folders? Could we, for instance, block creation? I am not arguing that we should do that, but wondering what is possible. I think it is at least _unexpected_ if I create a directory, and then find that it is pre-populated.

Reply:
I don't think we want to block the user from doing something that they may need to do.  Currently it will start projecting what should be in the folder and I think that is what would be most helpful to the user because if we
1. Block creation then the user is blocked if that is what they really need to do and we have no way to inform the user that they need to run a `gvfs include -a <folder>` in order to have that folder.
2. Allow them to create folders without showing the projection.  The user could create files that already exist in the index, in which case it would show as modified with only the content from the newly created file and they would have to run a `git checkout -- <file>` to get the content from the commit.

Starting the projection when the folder is created will let the user know the folder is currently a part of the repo and they will be able to see all the files that already exist in that folder, which to me would be more helpful in knowing what happened as soon as possible instead of later when I run a `git status` or something and wonder why.

Reply:
Thanks for the detailed response! I was mainly curious about what the possible options were. I think all options have the potential to cause confusion.

Is there any precedent from similar source control systems (if there is anything comparable...) that would point to what users would be familiar with?

I assume that creating new files outside the scope of "included" files will be a relatively (or at least not the common) scenario, and most likely indicates a deficit in the list of included directories? If I need to access files outside of this scope, I would expect to add the directory to the included list first.

If I run `mkdir`, it also might not be clear that there are new files because I happened to create a directory that collides with an non-included directory. There is no way to explain tothe user why the new entries are now present.

If the command was run as part of a script, it might hide the fact that the commands were running in an environment with insufficient include list entries.

Anyway - just my thoughts on what I *might* expect - I agree that blocking with an error is not a good experience either. Maybe just something for us to keep an eye out for as people use it.